src/files.js

```diff
@@ -290,5 +286,141 @@ module.exports = (common) => {
       })
     })
+
+    describe('.get', () => {
+      it('with a base58 encoded multihash', (done) => {
+        const hash = 'QmPZ9gcCEpqKTo6aq61g2nXGUhM4iCL3ewB6LDXZCtioEB'
+        ipfs.files.get(hash, (err, stream) => {
+          expect(err).to.not.exist
+          stream.pipe(concat((files) => {
+            expect(err).to.not.exist
+            expect(files).to.be.length(1)
+            expect(files[0].path).to.equal(hash)
+            files[0].content.pipe(concat((content) => {
+              expect(content.toString()).to.contain('Check out some of the other files in this directory:')
+              done()
+            }))
+          }))
+        })
+      })
+
+      it('with a multihash', (done) => {
+        const hash = 'QmPZ9gcCEpqKTo6aq61g2nXGUhM4iCL3ewB6LDXZCtioEB'
+        const mhBuf = new Buffer(bs58.decode(hash))
+        ipfs.files.get(mhBuf, (err, stream) => {
+          expect(err).to.not.exist
+          stream.pipe(concat((files) => {
+            expect(files).to.be.length(1)
+            expect(files[0].path).to.deep.equal(hash)
+            files[0].content.pipe(concat((content) => {
+              expect(content.toString()).to.contain('Check out some of the other files in this directory:')
+              done()
+            }))
+          }))
+        })
+      })
+
+      it('large file', (done) => {
+        const hash = 'Qme79tX2bViL26vNjPsF3DP1R9rMKMvnPYJiKTTKPrXJjq'
+        ipfs.files.get(hash, (err, stream) => {
+          expect(err).to.not.exist
+
+          // accumulate the files and their content
+          var files = []
+          stream.pipe(through.obj((file, enc, next) => {
+            file.content.pipe(concat((content) => {
+              files.push({
+                path: file.path,
+                content: content
+              })
+              next()
+            }))
+          }, () => {
+            expect(files.length).to.equal(1)
+            expect(files[0].path).to.equal(hash)
+            expect(files[0].content).to.deep.equal(bigFile)
+            done()
+          }))
+        })
+      })
+
+      it('directory', (done) => {
+        const hash = 'QmVvjDy7yF7hdnqE8Hrf4MHo5ABDtb5AbX6hWbD3Y42bXP'
+        ipfs.files.get(hash, (err, stream) => {
+          expect(err).to.not.exist
+
+          // accumulate the files and their content
+          var files = []
+          stream.pipe(through.obj((file, enc, next) => {
+            if (file.content) {
+              file.content.pipe(concat((content) => {
+                files.push({
+                  path: file.path,
+                  content: content
+                })
+                next()
+              }))
+            } else {
+              files.push(file)
+              next()
+            }
+          }, () => {
+            var paths = files.map((file) => {
+              return file.path
+            })
+            expect(paths).to.deep.equal([
+              'QmVvjDy7yF7hdnqE8Hrf4MHo5ABDtb5AbX6hWbD3Y42bXP',
```

Comment:
I think since this is the Core API we have to worry less about what the user expects and more about what would be the most useful to implementation developers, since they'll be the ones who ultimately decide what the users see. They *could* choose to cull the root hash from the result if they wanted.

My thinking is that, as an IPFS implementation developer, I'd want the root hash when `get`ing a directory: if I didn't then I'd need to do the extra legwork of inferring it from the path of its contents in order to figure out what root directory to create on the user's filesystem (in the case of `ipfs get`). This seems like a reasonable expectation to me, since the output of `get` will always match the directory structure that gets created by it (on the CLI, at least). Please let me know if you have any strong reasons against, but I think if not then this is a strong enough reason for.